Compiler warnings in _iomodule.c

I'm getting a compiler warning for `_iomodule.c`:

```console
./Modules/_io/_iomodule.c:649:15: warning: incompatible function pointer types initializing 'freefunc' (aka 'void (*)(void *)') with an expression of type 'void (PyObject *)' (aka 'void (struct _object *)') [-Wincompatible-function-pointer-types]
    .m_free = iomodule_free,
              ^~~~~~~~~~~~~
```

<!-- gh-linked-prs -->
### Linked PRs
* gh-103618
<!-- /gh-linked-prs -->

--- a/Modules/_io/_iomodule.c
+++ b/Modules/_io/_iomodule.c
@@ -616,8 +616,9 @@ iomodule_clear(PyObject *mod) {
 }
 
 static void
-iomodule_free(PyObject *mod) {
-    iomodule_clear(mod);
+iomodule_free(void *mod)
+{
+    (void)iomodule_clear((PyObject *)mod);
 }
 
 

